### PR TITLE
Add per-container metrics

### LINKED ISSUE #2116: Metrics: container space usage
Need a new metric from storage nodes to collect container space usage.
If it is possible to indicate the owner of the container with a label in this metric, then this will simplify further use.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #2116, #2117, #2118.

### CODE REVIEW

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3761

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/2Saj7qz1zckgxuSA9umLq1bMTLdHtoaxShvRx3oEweJ3">link</a>. Untar and use `allure open` to watch the report

File: pkg/services/object/metrics.go [hunk 1]
```diff
@@ -161,8 +171,18 @@ func (s putStreamMetric) Send(req *object.PutRequest) error {
...
 
-	s.metrics.IncPutReqCounter(err == nil)
+	s.metrics.IncPutReqCounter(s.cid, err == nil)
 	s.metrics.AddPutReqDuration(time.Since(s.start))
 
 	return res, err
 }
+
+func getCIDString(cnrV2 *refs.ContainerID) string {
+	var cnr cidSDK.ID
```
Comment (carpawell) [inline pkg/services/object/metrics.go:189]: we already have **v2** struct here, it is quite often operation, mb just `base58.Encode` with comment here?

File: pkg/local_object_storage/shard/metrics_test.go [hunk 2]
```diff
@@ -16,7 +16,8 @@ import (
 )
 
 type metricsStore struct {
-	s map[string]uint64
+	s       map[string]uint64
```
Comment (carpawell) [inline pkg/local_object_storage/shard/metrics_test.go:19]: lets name it `objCounters` then

File: pkg/local_object_storage/metabase/delete.go [hunk 3]
```diff
@@ -144,7 +154,7 @@ func (db *DB) deleteGroup(tx *bbolt.Tx, addrs []oid.Address) (uint64, uint64, er
 // non-exist object is error-free). The second return value indicates if an
 // object was available before the removal (for calculating the logical object
 // counter).
```
Comment (carpawell) [inline pkg/local_object_storage/metabase/delete.go:156]: add the third val desc?

**Comment by carpawell** [comment]
CHANGELOG

File: pkg/services/object/metrics.go (hunk 1 above)
Comment (fyrchik) [reply pkg/services/object/metrics.go:189]: Fixed

File: pkg/local_object_storage/shard/metrics_test.go (hunk 2 above)
Comment (fyrchik) [reply pkg/local_object_storage/shard/metrics_test.go:19]: Fixed

File: pkg/local_object_storage/metabase/delete.go (hunk 3 above)
Comment (fyrchik) [reply pkg/local_object_storage/metabase/delete.go:156]: Fixed

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3764

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3765

**Comment by sami-nspcc** [comment]
I am running integration tests

**Review by carpawell — APPROVED** [review]

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/HFTuwJzFs5Gi3pDmxBLCNbv7faSZ8K12zTuoZY9HrDkw">link</a>. Untar and use `allure open` to watch the report

File: pkg/local_object_storage/shard/shard.go [hunk 4]
```diff
@@ -334,6 +337,23 @@ func (s *Shard) updateObjectCounter() {
...
+
+		for i := range cnrList {
+			size, err := s.metaBase.ContainerSize(cnrList[i])
+			if err != nil {
+				s.log.Warn("meta: can't read container size",
+					zap.String("cid", cnrList[i].EncodeToString()),
+					zap.Error(err))
+				continue
+			}
+			s.metricsWriter.AddToContainerSize(cnrList[i].EncodeToString(), int64(size))
```
Comment (cthulhu-rider) [inline pkg/local_object_storage/shard/shard.go:355]: Can `metricsWriter` be `nil` here?

Is this method called once? Maybe name it `initMetrics` then?

**Comment by fyrchik** [comment]
Moving this to draft until the final decision about when we should merge it.

**Pull request closed without merging**